File: csharp/demo/demo12b02-data-race-multi.cs
/*
 * DATA RACES
 * Version 02: Multithreading
 */
using System;
using System.Threading;
using System.Linq;



class Demo12B02 : IRunnable
{
    public void run()
    {
        const int N = 8;
        var a = Enumerable.Repeat(false, N + 1).ToArray();


        var thDiv2 = new Thread(() =>
        {
            for (int i = 2; i <= N; i += 2)
                a[i] = true;
        });

        var thDiv3 = new Thread(() =>
        {
            for (int i = 3; i <= N; i += 3)
                a[i] = true;
        });


        thDiv2.Start();
        thDiv3.Start();
        thDiv2.Join();
        thDiv3.Join();


        int result = a.Count(val => val);
        Console.WriteLine("Numbers of integers that are divisible by 2 or 3 is: " + result);
    }
}

File: python/demo12b02_data_race_multi.py
'''
DATA RACES
Version 02: Multithreading
'''

import threading



def count_div_2(a: list, n: int):
    for i in range(2, n + 1, 2):
        a[i] = True



def count_div_3(a: list, n: int):
    for i in range(3, n + 1, 3):
        a[i] = True



N = 8
A = [False] * (N + 1)

th_div_2 = threading.Thread(target=count_div_2, args=(A, N))
th_div_3 = threading.Thread(target=count_div_3, args=(A, N))

th_div_2.start()
th_div_3.start()
th_div_2.join()
th_div_3.join()

result = A.count(True)

print('Number of integers that are divisible by 2 or 3 is:', result)
